packages/core/workers/src/Handle.js

```diff
@@ -1,17 +1,43 @@
-import WorkerFarm from './WorkerFarm';
-import packageJson from '../package.json';
+// @flow strict-local
+
+import type {WorkerApi} from './';
+
 import {registerSerializableClass} from '@parcel/utils';
+import nullthrows from 'nullthrows';
+
+import {child} from './childState';
+// $FlowFixMe this is untyped
+import packageJson from '../package.json';
 
 let HANDLE_ID = 0;
+let handleIdToWorkerApi = new Map();
+
+type HandleOpts = {|
+  id?: number,
+  workerApi?: WorkerApi
+|};
 
 export default class Handle {
-  constructor(opts) {
-    this.id = opts ? opts.id : ++HANDLE_ID;
+  id: number;
+
+  constructor(opts: HandleOpts) {
+    this.id = opts?.id ?? ++HANDLE_ID;
+    if (opts?.workerApi) {
+      handleIdToWorkerApi.set(this.id, opts.workerApi);
+    }
   }
 
-  static deserialize(opts) {
-    return function(...args) {
-      return WorkerFarm.callMaster({handle: opts.id, args}, true);
+  dispose() {
+    handleIdToWorkerApi.delete(this.id);
+  }
+
+  static deserialize(opts: {|id: number|}) {
+    return function(...args: Array<mixed>) {
+      let workerApi = child
+        ? child.workerApi
+        : nullthrows(handleIdToWorkerApi.get(opts.id));
```

Comment:
when does this case happen, i.e. when is there no child state?

Reply:
This can happen any time the handle is executed in the main process — like in the local worker.